quinn-proto/src/token.rs

```diff
@@ -99,19 +93,36 @@ fn decode_addr<B: Buf>(buf: &mut B) -> Option<SocketAddr> {
     Some(SocketAddr::new(ip, port))
 }
 
-/// Reasons why a retry token might fail to validate a client's address
+/// Error for a token failing to validate a client's address
 #[derive(Debug, Copy, Clone)]
-pub(crate) enum TokenDecodeError {
-    /// Token was not recognized. It should be silently ignored.
-    UnknownToken,
-    /// Token was well-formed but associated with an incorrect address. The connection cannot be
-    /// established.
-    WrongAddress,
+pub(crate) enum ValidationError {
+    /// Token may have come from a NEW_TOKEN frame (including from a different server or a previous
+    /// run of this server with different keys), and was not valid
+    ///
+    /// It should be silently ignored.
+    ///
+    /// In cases where a token cannot be decrypted/decoded, we must allow for the possibility that
+    /// this is caused not by client malfeasance, but by the token having been generated by an
+    /// incompatible endpoint, e.g. a different version or a neighbor behind the same load
+    /// balancer. In such cases we proceed as if there was no token.
+    ///
+    /// [_RFC 9000 § 8.1.3:_](https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1.3-10)
+    ///
+    /// > If the token is invalid, then the server SHOULD proceed as if the client did not have a
+    /// > validated address, including potentially sending a Retry packet.
+    ///
+    /// That said, this may also be used when a token _can_ be unambiguously decrypted/decoded as a
+    /// token from a NEW_TOKEN frame, but is simply not valid.
+    Ignore,
```

Comment:
nit: It feels weird for this to be a verb, when the other variant is (and error variants in general are) a noun. I like `Unknown` a little more in that respect, though I'm open to other ideas.

Reply:
Yeah, it's a bit unfortunate. I touched on this above, but to recap: "Unknown" has the problem of, "why do we return Unknown for a token from NEW_TOKEN that's expired or has the wrong address, we clearly know what this token is?", and something both nouny and accurate would have to capture "this is not valid, and it could conceivably be not from a Retry," which is cumbersome.

Reply:
Maybe `Ignored`?

Reply:
Maybe, but if I saw `Ignored` decontextualized, I would imagine it to mean "the problem is that this has been ignored" before I'd imagine it to mean 'the error is that this thing should be ignored", or something

Reply:
Maybe `Unusable`?

Reply:
I kind of like `Unusable`...

Reply:
Works for me!